File: src/aim.cpp
#include "cogstate/aim.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace cogstate {

AdaptiveInheritedMasking::AdaptiveInheritedMasking(AimConfig config) : config_(config) {
  if (config_.inherited_decay <= 0.0 || config_.inherited_decay > 1.0) {
    throw std::runtime_error("inherited_decay must be in (0, 1]");
  }
}

AimOutput AdaptiveInheritedMasking::apply(const Matrix& signal,
                                          const std::vector<std::vector<bool>>& observed) const {
  if (signal.size() != observed.size()) {
    throw std::runtime_error("signal/observed time dimension mismatch");
  }
  if (signal.empty()) {
    return {};
  }

  const std::size_t t_len = signal.size();
  const std::size_t dim = signal.front().size();
  for (std::size_t t = 0; t < t_len; ++t) {
    if (signal[t].size() != dim || observed[t].size() != dim) {
      throw std::runtime_error("signal/observed feature dimension mismatch");
    }
  }

  Matrix imputed = signal;
  std::vector<std::vector<double>> weights(t_len, std::vector<double>(dim, 1.0));
  std::vector<double> last_value(dim, 0.0);
  std::vector<int> missing_run(dim, 0);
  std::vector<bool> initialized(dim, false);

  for (std::size_t t = 0; t < t_len; ++t) {
    for (std::size_t d = 0; d < dim; ++d) {
      if (observed[t][d]) {
        initialized[d] = true;
        last_value[d] = signal[t][d];
        missing_run[d] = 0;
        weights[t][d] = 1.0;
      } else {
        missing_run[d] += 1;
        const double decay = std::pow(config_.inherited_decay, static_cast<double>(missing_run[d]));
        const bool nmar_like = missing_run[d] >= 3;
        const double boost = nmar_like ? config_.nmar_run_boost : 1.0;
        weights[t][d] = std::min(2.0, boost * decay);
        if (initialized[d]) {
          imputed[t][d] = last_value[d] * weights[t][d];
        } else {
          imputed[t][d] = 0.0;
        }
      }
    }
  }

  return {imputed, weights};
}

}  // namespace cogstate

File: ts/aim/aim.ts
export interface AimConfig {
  inheritedDecay: number;
  nmarRunBoost: number;
}

export interface AimOutput {
  imputed: number[][];
  adaptiveWeights: number[][];
}

export class AdaptiveInheritedMasking {
  constructor(private readonly cfg: AimConfig) {
    if (!(cfg.inheritedDecay > 0 && cfg.inheritedDecay <= 1)) {
      throw new Error("inheritedDecay must be in (0,1]");
    }
  }

  apply(signal: number[][], observed: boolean[][]): AimOutput {
    if (signal.length !== observed.length) throw new Error("signal/observed length mismatch");
    if (signal.length === 0) return { imputed: [], adaptiveWeights: [] };

    const tLen = signal.length;
    const dim = signal[0].length;
    const imputed = signal.map((x) => [...x]);
    const adaptiveWeights = Array.from({ length: tLen }, () => new Array<number>(dim).fill(1));
    const lastValue = new Array<number>(dim).fill(0);
    const missingRun = new Array<number>(dim).fill(0);
    const initialized = new Array<boolean>(dim).fill(false);

    for (let t = 0; t < tLen; t += 1) {
      for (let d = 0; d < dim; d += 1) {
        if (observed[t][d]) {
          initialized[d] = true;
          lastValue[d] = signal[t][d];
          missingRun[d] = 0;
          adaptiveWeights[t][d] = 1;
        } else {
          missingRun[d] += 1;
          const decay = this.cfg.inheritedDecay ** missingRun[d];
          const boost = missingRun[d] >= 3 ? this.cfg.nmarRunBoost : 1;
          const w = Math.min(2, boost * decay);
          adaptiveWeights[t][d] = w;
          imputed[t][d] = initialized[d] ? lastValue[d] * w : 0;
        }
      }
    }

    return { imputed, adaptiveWeights };
  }
}

